engine core

victorbalan opened the pull request:
- Added /api/start/simple route wich starts a run with the received code and language
- Added support for c, c++, python2.7 and Java
- Added dockerfile for creating a special container containing all the required compilers

victorbalan committed:
Fix windows dir generation

victorbalan committed:
engine core

 * Added /api/start/simple route wich starts a run with the received code and language
 * Added support for c, c++, python2.7 and Java

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		return
+	}
+	dst = dst + "/" + fileName
```
Use [`path.Join`](http://golang.org/pkg/path/#Join) for cross-platform compatibility.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+		return
+	}
+	dst = dst + "/" + fileName
+	defer in.Close()
```
Move this `defer` right after the error check.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
```
This will set `err` to `nil` if `out.Close()` succeeds, effectively overwriting previous errors. Is that intended?

lorenzleutgeb replied:
Sorry, I was mistaken.

``` go
defer func() {
    if cerr := out.Close(); err == nil {
        err = cerr
    }
}
```

would me more concise, though.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+		dockerHomeDir = strings.Replace(dockerHomeDir, "C", "c", 1)
+		dockerHomeDir = strings.Replace(dockerHomeDir, "\\", "/", -1)
+	} else {
+		homeDir = os.Getenv("HOME") + "/tmp/"
```
This is a no-go. We won't just fuck around with our `$HOME`s. Use [`ioutil.TempDir`](https://golang.org/pkg/io/ioutil/#TempDir) in in that case.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+			fmt.Println("Your home path should begin with C:\\Users\\ because you are on windows and that`s the only boot2docker shared folder")
+			return
+		}
+		dockerHomeDir = "/" + homeDir
```
Use the previous assertion (`strings.HasPrefix`) to avoid calling `strings.Replace` more than once by indexing:

``` go
if !strings.HasPrefix(u.HomeDir, "C:\\Users\\") {
    ...
    return
}
dockerHomedir = "/c" + strings.Replace(u.HomeDir[2:],  `\`, `/`, -1)
```

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+			return
+		}
+		homeDir = u.HomeDir
+		homeDir = strings.Replace(homeDir, "\\", "/", -1) + "/tmp/"
```
Not sure why this is here. Move it below the following if or delete it.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+	// keep both home dir and docker home dir even if they are the same because
+	// on windows we need to have a dockerHomeDir=/c/Users/$USER/tmp and a
+	// homeDir=C:/Users/%USER%/tmp. On Linux they are homeDir=dockerHomeDir=$HOME/tmp
+	homeDir       string
```
I do not see why we store the user's home directory here. We should rather focus on providing two different kinds of _temporary directories_. Once for configurations, and one acting as Docker volume.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+		"--rm",
+		"-v",
+		dockerTempDir+":/run",
+		"coduno_all")
```
OT: Image names usually use hyphens instead of underlines.

lorenzleutgeb commented on engine.go:
```diff
@@ -0,0 +1,247 @@
+	if _, err = io.Copy(out, in); err != nil {
+		return
+	}
+	err = out.Sync()
```
Why do you force a sync here?

lorenzleutgeb committed:
Improve engine core

 * Transparent "dockerization" of paths due to path mangling
   needed for mounting Docker volumes on Windows with boot2docker.
 * Nicer creation of temporary directories (got rid of silly UUID
   dependency).
 * Code cleanup (especially the clutter in main.main)
 * Not forcing the FS to sync upon copy anymore
 * Use the same parameter order for copyFileContents as io.Copy as
   it is just a higher level version of it.
